effectful/internals/base_impl.py

```diff
@@ -51,13 +32,20 @@ def __hash__(self):
         return hash(self.signature)
 
     def __default_rule__(self, *args: Q.args, **kwargs: Q.kwargs) -> "Expr[V]":
+        from effectful.ops.syntax import defdata
+
         try:
             return self.signature(*args, **kwargs)
         except NotImplementedError:
-            return self.__free_rule__(*args, **kwargs)
+            return typing.cast(
+                Callable[Concatenate[Operation[Q, V], Q], Expr[V]], defdata
+            )(self, *args, **kwargs)
 
-    def __free_rule__(self, *args: Q.args, **kwargs: Q.kwargs) -> "Expr[V]":
-        from effectful.ops.syntax import Bound, Scoped, defdata, defop
+    def __fvs_rule__(self, *args: Q.args, **kwargs: Q.kwargs) -> tuple[
+        tuple[collections.abc.Set[Operation], ...],
+        dict[str, collections.abc.Set[Operation]],
+    ]:
```

Comment:
Let's add some documentation here. It's unclear to me exactly what this rule does now.

Reply:
As of this PR, `Operation.__fvs_rule__` is supposed to return the variables appearing free in the scope of each positional and keyword argument that are newly bound by the `Operation`. The results are aggregated by `fvsof` to compute the free variables of a `Term`.  For example:
```py
@defop
def Let(
  var: Annotated[Operation[[], S], Bound(0)],
  val: Annotated[S, Scoped(1)],
  body: Annotated[T, Scoped(0)],
) -> T: ...

x, y = defop(int), defop(int)
val = y() + 1
body = x() + y() + 2
assert Let.__fvs_rule__(x, val, body) == ({x}, set(), {x})
assert fvsof(Let(x, val, body)) == {y}
```

Reply:
Should the first assert be `Let.__fvs_rule__(x, val, body) == (({x}, set(), {x}), {})`?

Reply:
I see. Is the idea that these are the variables that are safe to alpha-rename? I think including this description in the docstring is a good idea, since this is different from a standard free variables computation.

Reply:
In addressing #171 in a followup #208 to this PR, I've made further significant changes to binding. I think it makes more sense to update the documentation there, which I've done a first pass on.